### PR TITLE
Files upload/download utils

### PR DESCRIPTION
**Pull request by Starzu**
You can find usage demo here: https://github.com/UdashFramework/udash-demos/tree/files-upload/file-upload

### CODE REVIEW

File: core/frontend/src/main/scala/io/udash/utils/FileUploader.scala [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  def upload(fieldName: String, files: Seq[File],
+             extraData: Map[js.Any, js.Any] = Map.empty): ReadableModelProperty[FileUploadModel] = {
+    val p = ModelProperty[FileUploadModel]
+    val data = new FormData()
+
+    extraData.foreach(item => data.append(item._1, item._2))
+    files.foreach(file => {
+      data.append(s"$fieldName[]", file)
```
Comment (grzesiul) [inline core/frontend/src/main/scala/io/udash/utils/FileUploader.scala]: Why is this "[ ]" required? Does this connect with FileUploadServlet:17 - it's stripped there.

File: rpc/backend/src/main/scala/io/udash/rpc/utils/FileDownloadServlet.scala [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    * By default returns the original file name. */
+  protected def presentedFileName(name: String): String = name
+
+  override def doGet(request: HttpServletRequest, response: HttpServletResponse): Unit = {
+    val file = resolveFile(request)
+
+    if (!file.exists()) response.sendError(404, "File not found!")
+    else {
+      val context = getServletContext
+      val mimeType = Option(context.getMimeType(file.getAbsolutePath)).getOrElse("application/octet-stream")
```
Comment (grzesiul) [inline rpc/backend/src/main/scala/io/udash/rpc/utils/FileDownloadServlet.scala]: I'd make from that 'resolveMimeType' with provided default implementation. One may want to enforce mime type or 'manually' resolve it.

File: rpc/backend/src/main/scala/io/udash/rpc/utils/FileDownloadServlet.scala [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+      response.setContentType(mimeType)
+      if (file.length() <= Int.MaxValue) response.setContentLength(file.length.toInt)
```
Comment (grzesiul) [inline rpc/backend/src/main/scala/io/udash/rpc/utils/FileDownloadServlet.scala]: What if file length > Int.MaxValue, can response has null contentLength ?

File: rpc/backend/src/main/scala/io/udash/rpc/utils/FileUploadServlet.scala [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  * @param fileFields Names of file inputs in the HTTP request. */
+abstract class FileUploadServlet(fileFields: Set[String]) extends HttpServlet {
+  /** Uploaded file handler. */
+  protected def handleFile(name: String, content: InputStream): Unit
+
+  override protected def doPost(request: HttpServletRequest, response: HttpServletResponse): Unit = {
+    import scala.collection.JavaConversions._
+
+    request.getParts.toStream
+      .filter(part => fileFields.contains(part.getName.stripSuffix("[]")))
```
Comment (grzesiul) [inline rpc/backend/src/main/scala/io/udash/rpc/utils/FileUploadServlet.scala]: What's "[ ]" used for?

File: rpc/backend/src/main/scala/io/udash/rpc/utils/FileDownloadServlet.scala (hunk 3 above)
Comment (ddworak) [inline rpc/backend/src/main/scala/io/udash/rpc/utils/FileDownloadServlet.scala]: http://stackoverflow.com/questions/11466905/how-to-set-content-length-as-long-value-in-http-header-in-java

File: core/frontend/src/main/scala/io/udash/utils/FileUploader.scala [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+    upload(input.name, builder.result())
+  }
+
+  def upload(fieldName: String, files: Seq[File],
+             extraData: Map[js.Any, js.Any] = Map.empty): ReadableModelProperty[FileUploadModel] = {
+    val p = ModelProperty[FileUploadModel]
+    val data = new FormData()
+
+    extraData.foreach(item => data.append(item._1, item._2))
```
Comment (ddworak) [inline core/frontend/src/main/scala/io/udash/utils/FileUploader.scala]: Use case and extract tuple elements with some nice names/

File: core/frontend/src/main/scala/io/udash/utils/FileUploader.scala [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+import org.scalajs.dom._
+
+import scala.concurrent.ExecutionContext
+import scala.scalajs.js
+
+class FileUploader(url: Url)(implicit ec: ExecutionContext) {
+  import FileUploader._
+
+  def upload(input: html.Input): ReadableModelProperty[FileUploadModel] = {
+    val builder = Seq.newBuilder[File]
```
Comment (ddworak) [inline core/frontend/src/main/scala/io/udash/utils/FileUploader.scala]: If performance is not an issue, you could just
```scala
upload(input.name, (0 until input.files.length).map(i => input.files(i)))
```

File: core/frontend/src/main/scala/io/udash/utils/FileUploader.scala (hunk 1 above)
Comment (Starzu) [inline core/frontend/src/main/scala/io/udash/utils/FileUploader.scala]: It means there might be more than one file.

**Comment by Starzu** [comment]
@grzesiul @ddworak take a look again, please.

**Comment by grzesiul** [comment]
lgtm

**Comment by ddworak** [comment]
lgtm

**Pull request merged**